### PR TITLE
Added Results Scraping after a suite is run

### PR DESCRIPTION
**Pull request by grapheo12**
After the usual run of `teuthology-results`, the scraper script produces a summary of the results and stores it in `scrape.log` inside the `archive-dir` of the results.

## Changes made

- Added the scraper script from https://github.com/jcsp/scrape as `teuthology/scrape.py`.

- Modified the script at line 470 to include printing of the scrape logs.

- Modified `teuthology/results.py` to run the scraper after every result generation and emailing.

## Testing

Since I don't have access to a paddle database, I did the following to test:

1. Installed the package using `pip install .`.

2. In the site-packages, commented out / filled with dummy data, the lines that required calling the endpoints on the database.

3. Then ran:
```sh
teuthology-results --archive-dir results --name rados --email a@b.com --dry-run --timeout 2
```
Here results is a sample results directory given by @jdurgin 

Signed-off-by: Shubham Mishra <smishra99.iitkgp@gmail.com>

### CODE REVIEW

File: teuthology/results.py [hunk 1]
```diff
@@ -82,6 +83,7 @@ def results(archive_dir, name, email, timeout, dry_run):
             )
     finally:
         generate_coverage(archive_dir, name)
+        Scraper(archive_dir).analyze()
```
Comment (jdurgin) [inline teuthology/results.py]: to ensure this gets run, even if generate_coverage() fails, you could wrap generate_coverage in a try/finally as well

File: teuthology/scrape.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
+#!/usr/bin/env python
+
```
Comment (jdurgin) [inline teuthology/scrape.py]: since we're importing this from another repo, let's add a comment here about its origin and original author. Then add your additional changes to it as another commit, so it's easy to tell what you added

**Comment by jdurgin** [comment]
jenkins: ok to test

File: teuthology/scrape.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
+#!/usr/bin/env python
```
Comment (jdurgin) [inline teuthology/scrape.py]: don't need the #! line since we're importing it, not running it as a script now

**Comment by grapheo12** [comment]
Sorry, I wrongly ran git rebase with HEAD~4. That is why previous history of other commits is coming here.

**Comment by grapheo12** [comment]
@jdurgin Please review.

**Comment by tchaikov** [comment]
could you add a "`Signed-off-by`" line at the end of your commit message? "git commit -s" will do the trick for you. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work

**Comment by grapheo12** [comment]
> could you add a "`Signed-off-by`" line at the end of your commit message? "git commit -s" will do the trick for you. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work

Added manually for now. Will keep in mind from now on. Thanks

**Comment by jdurgin** [comment]
@grapheo12 please add it to the commit messages - you can do this by interactive rebase, e.g. ```git rebase -i origin/master```, mark both commits to be edited, and for each one ```git commit --amend -s && git rebase --continue```

**Comment by grapheo12** [comment]
@jdurgin Done.

**Comment by jdurgin** [comment]
ok to test

**Comment by kshtsk** [comment]
ok to test

File: teuthology/scrape.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+    """
+    A known reason matching a particular regex to failure reason
+    """
+
+    def __init__(self, regexes, description):
+        self.description = description
+        if not isinstance(regexes, list):
+            self.regexes = [regexes]
+        else:
+            self.regexes = regexes
```
Comment (kshtsk) [inline teuthology/scrape.py]: Can this have change check order to avoid negative condition in if for straight logic.

File: teuthology/scrape.py [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+            return timeout, command
+        else:
+            return None
+
+    def match(self, job):
+        if self.get_timeout(job) == (self.timeout, self.command):
+            return True
+        else:
+            return False
```
Comment (kshtsk) [inline teuthology/scrape.py]: this can be written as a single line:
```
return self.get_timeout(job) == (self.timeout, self.command)
```

File: teuthology/scrape.py (hunk 4 above)
Comment (grapheo12) [reply teuthology/scrape.py]: Do you mean just flipping the lines of if and else?

File: teuthology/scrape.py [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+    def could_be(cls, job):
+        return cls.get_timeout(job) is not None
+
+    @classmethod
+    def get_timeout(cls, job):
+        if job.get_failure_reason() is None:
+            return None
+
+        match = re.search("status 124:.* timeout ([^ ]+) ([^']+)'", job.get_failure_reason())
+        if match:
```
Comment (kshtsk) [inline teuthology/scrape.py]: It would be better to avoid extra indentation level like:
```
if not match:
    return None
.
.
.
```

File: teuthology/scrape.py [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+                return ratio > 0.5
+            else:
+                # I have BT but he doesn't, so we're different
+                return False
+
+        if self.last_tlog_line is None and job.get_last_tlog_line() is None:
+            return True
+
+        ratio = difflib.SequenceMatcher(None, self.last_tlog_line, job.get_last_tlog_line()).ratio()
+        return ratio > 0.5
```
Comment (kshtsk) [inline teuthology/scrape.py]: For better readability this can be rewritten like:
```
if self.last_tlog_line or job.get_last_tlog_line():
          ratio = difflib.SequenceMatcher(None, self.last_tlog_line,
                                          job.get_last_tlog_line()).ratio()
          return ratio > 0.5
return True
```

File: teuthology/scrape.py [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+
+    def get_detail(self):
+        return self.backtrace
+
+    @classmethod
+    def could_be(cls, job):
+        return job.summary_data is None
+
+    def match(self, job):
+        if job.summary_data is not None:
```
Comment (kshtsk) [inline teuthology/scrape.py]: that should be more pythonish:
```
if job.summary_data:
    return False
```

File: teuthology/scrape.py [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+    """
+    @classmethod
+    def could_be(cls, job):
+        if not super(LockdepReason, cls).could_be(job):
+            return False
+
+        if "common/lockdep" in job.get_assertion():
+            return True
+        else:
+            return False
```
Comment (kshtsk) [inline teuthology/scrape.py]: That can be simply written as single line:
```
return "common/lockdep" in job.get_assertion()
```

File: teuthology/scrape.py (hunk 4 above)
Comment (kshtsk) [reply teuthology/scrape.py]: Right, like:
```
if isinstance(regexes, list):
    self.regexes = regexes
else:
    self.regexes = [regexes]
```

**Comment by kshtsk** [comment]
I wonder if this code has a unit tests?

**Comment by jdurgin** [comment]
@kshtsk the goal for this PR is just to get scrape into the repo - it's been quite helpful but many users of teuthology are unaware of it - we can continue improving it in other PRs. I'm happy merging it as-is and improving tests and style in further PRs. Is that ok with you?

File: teuthology/scrape.py [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+
+
+def grep(path, expr):
+    """
+    Call out to native grep rather than feeding massive log files through python line by line
+    """
+    p = subprocess.Popen(["grep", expr, path], stdout=subprocess.PIPE)
+    out, err = p.communicate()
+    if p.returncode == 0:
+        return out.split("\n")
```
Comment (kshtsk) [inline teuthology/scrape.py]: Not sure this code is py3 compatible.

File: teuthology/scrape.py [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+        else:
+            self.regexes = [regexes]
+
+    def match(self, job):
+        # I never match dead jobs
+        if job.get_failure_reason() is None:
+            return False
+
+        for regex in self.regexes:
+            if re.match(regex, job.get_failure_reason()) is not None:
```
Comment (kshtsk) [inline teuthology/scrape.py]: `is not None` is extra here

**Comment by kshtsk** [comment]
> @kshtsk the goal for this PR is just to get scrape into the repo - it's been quite helpful but many users of teuthology are unaware of it - we can continue improving it in other PRs. I'm happy merging it as-is and improving tests and style in further PRs. Is that ok with you?

it is okay as soon as it is py2/py3 compatible

**Comment by kshtsk** [comment]
```
E     File "/home/jenkins-build/build/workspace/teuthology-pull-requests/teuthology/scrape.py", line 147
E       return "common/lockdep" in job.get_assertion():
E                                                     ^
E   SyntaxError: invalid syntax
```

File: teuthology/scrape.py (hunk 10 above)
Comment (grapheo12) [reply teuthology/scrape.py]: Adding a `p.wait()` sets `p.returncode` to 0 (if grep process was successful). Tested this in both py2 and py3. Otherwise line 34 evaluates to `True` and `False` in py2 and py3 respectively.

**Comment by grapheo12** [comment]
@kshtsk @jdurgin Please review.

File: teuthology/scrape.py (hunk 10 above)
Comment (kshtsk) [reply teuthology/scrape.py]: @grapheo12 the question is about `out.split("\n")`, since Popen will return Bytes in out but in python3 str should be prepended with `b` to compare with bytes.

**Comment by kshtsk** [comment]
@jdurgin is this code supposed to be running for each teuthology job?

File: teuthology/scrape.py (hunk 10 above)
Comment (grapheo12) [reply teuthology/scrape.py]: In that case, let me introduce a check:

```python
if type(out) != str:
    out = out.decode()
```

This should take care of the bytes vs str problem.

**Comment by jdurgin** [comment]
> @jdurgin is this code supposed to be running for each teuthology job?

Just once for each run, it checks all jobs in a run at once.

File: teuthology/scrape.py (hunk 10 above)
Comment (kshtsk) [reply teuthology/scrape.py]: We have a uniform practice for this kind of check to use `six.ensure_str`.

**Comment by kshtsk** [comment]
Well then this code for sure will break py3 based teuthology setup, we've recently merged last portion of py3 compatible code, and we cannot switch our sepia to py3 yet know, so we enable py3 unit/functional test for teuthology, so it is highly desired any new code have a test covering real py execution, because static analysis tools does not catch issues like I mentioned in a review.

**Comment by grapheo12** [comment]
If it is needed, I can write unit tests for it. Please guide on what kinds of checks are required.

**Comment by kshtsk** [comment]
> If it is needed, I can write unit tests for it. Please guide on what kinds of checks are required.

we basically need a functional test, since the code involve file operations, and system grep are involved, it is probably makes sense to have some shorten variant of real logs would be great to have parsed.

**Comment by grapheo12** [comment]
Should I do that on a separate PR?

File: teuthology/scrape.py [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+
+
+def grep(path, expr):
+    """
+    Call out to native grep rather than feeding massive log files through python line by line
+    """
+    p = subprocess.Popen(["grep", expr, path], stdout=subprocess.PIPE)
+    p.wait()
+    out, err = p.communicate()
+    out = six.ensure_str(out)
```
Comment (kshtsk) [inline teuthology/scrape.py]: Sorry for asking this, but self modifying variables is not a good practice, can you just put this:
```
if p.returncode:
    return []
else:
    return six.ensure_str(out).split("\n")
```

Comment (grapheo12) [reply teuthology/scrape.py]: I'll keep in mind from the next time. Fixing this right away.

**Comment by kshtsk** [comment]
> Should I do that on a separate PR?

It would be highly appreciated to have it within this PR, as I said, it will break py3 part if we not test it and merge.

File: teuthology/results.py [hunk 13]
```diff
@@ -81,7 +82,10 @@ def results(archive_dir, name, email, timeout, dry_run):
                 body=body,
             )
     finally:
-        generate_coverage(archive_dir, name)
+        try:
+            generate_coverage(archive_dir, name)
+        finally:
+            Scraper(archive_dir).analyze()
```
Comment (kshtsk) [inline teuthology/results.py]: @jdurgin @grapheo12 I'm actually okay to merge it early if you revert these, line and submit integration part of scrape within another PR with test covering scrape.

**Comment by kshtsk** [comment]
@susebot run deploy

File: teuthology/results.py (hunk 13 above)
Comment (grapheo12) [reply teuthology/results.py]: May I just comment this line and add ```pass``` statement?
Its all up to you. This is my first contribution to Ceph. :sweat_smile:
@kshtsk @jdurgin

**Comment by vasukulkarni** [comment]
Not blocking this PR, but please add a follow on PR to document this as well. Thanks.

File: teuthology/results.py (hunk 13 above)
Comment (kshtsk) [reply teuthology/results.py:88]: @grapheo12 Makes sense, also add a comment that this code is not tested and can be "enabled" after adding tests.

**Comment by grapheo12** [comment]
@vasukulkarni @kshtsk I am disabling the integration for now.
Should I put the unit tests, integration part and the documentation together in the next PR?

**Comment by vasukulkarni** [comment]
> @vasukulkarni @kshtsk I am disabling the integration for now.
> Should I put the unit tests, integration part and the documentation together in the next PR?

Sure that works!

**Comment by kshtsk** [comment]
@grapheo12 do you still plan to add tests as a separate PR or anything like that?

**Comment by grapheo12** [comment]
@kshtsk Yes I do plan to work on that. But due to my studies, I will not be able to concentrate on it till the 1st week of May. I'll definitely work on it after that. Please give me some time if that is not a problem.

**Comment by kshtsk** [comment]
> @kshtsk Yes I do plan to work on that. But due to my studies, I will not be able to concentrate on it till the 1st week of May. I'll definitely work on it after that. Please give me some time if that is not a problem.

np, take your time, it was just a heads-up

**Comment by kshtsk** [comment]
A month passed from the 1 May, there is a conflict already, also the master branch is switched to py3. So we want this merged, we need to:
1. resolve conflicts
2. add tests so we can cover py2 and py3 testing coverage so we can back port this to py2 as well.

**Comment by grapheo12** [comment]
@kshtsk Sorry for the delay. I started looking into it a week or so back. But could not decide what kind of tests to write for it. It'd help if you can provide me the with a brief overview of the test functions I need to write. I have no background in writing tests. Sorry again.

**Comment by grapheo12** [comment]
The code is compatible with both py2 and py3 I guess, based on the changes I made after the comments I received in this thread.

**Comment by grapheo12** [comment]
@kshtsk @jdurgin Please review if I am going in the right direction in writing the test.

**Comment by tchaikov** [comment]
retest this please.

**Comment by kshtsk** [comment]
hey @grapheo12 I'm sorry for so long response, however, I tried to run tests and can see messages like that:
```  /home/kyr/kshtsk/teuthology/teuthology/scrape.py:262: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
    self.config = yaml.load(open(os.path.join(self.path, "config.yaml"), 'r'))
  /home/kyr/kshtsk/teuthology/teuthology/scrape.py:271: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
    self.summary_data = yaml.load(open(summary_path, 'r'))
```
Can we get rid of it.

File: teuthology/test/test_scrape.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+        return self
+
+    def __exit__(self, exc_typ, exc_val, exc_tb):
+        shutil.rmtree(self.path)
+
+class TestScrape(object):
+    """Tests for teuthology.scrape"""
+
+    def test_grep(self):
+        with open("/tmp/scrapetest.txt", 'w') as f:
```
Comment (kshtsk) [inline teuthology/test/test_scrape.py:48]: The hardcoded file name is not thread safe, could you please use `tempfile` for creating temporary files?

**Comment by kshtsk** [comment]
@grapheo12 and I guess we can try and enable it.

File: teuthology/test/test_scrape.py (hunk 14 above)
Comment (grapheo12) [reply teuthology/test/test_scrape.py:48]: Sure. Doing this right away!

**Comment by grapheo12** [comment]
I am also integrating the upstream changes to the script with this.

File: teuthology/test/test_scrape.py (hunk 14 above)
Comment (grapheo12) [reply teuthology/test/test_scrape.py:48]: Actually `grep` takes a filename as a parameter. With `tempfile` I don't think you get a valid path.
So, I am thinking of using something like `uuid` as the filename.

Comment (grapheo12) [reply teuthology/test/test_scrape.py:48]: Used the `FakeResultDir` class I defined in the file.

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by kshtsk** [comment]
https://ceph-ci.suse.de/job/teuthology-deploy-ovh/269/artifact/logs/run/scrape.log/*view*/
Here is the log it generated.

**Comment by kshtsk** [comment]
Corresponding result.log:
https://ceph-ci.suse.de/job/teuthology-deploy-ovh/269/artifact/logs/run/results.log/*view*/

**Comment by kshtsk** [comment]
@grapheo12 I'm sorry I removed your comment with that big diff since it is not relevant and present a noise.

**Comment by kshtsk** [comment]
I cleaned the DNM tag.

**Comment by kshtsk** [comment]
@grapheo12 maybe you can squash the commits?

**Comment by grapheo12** [comment]
@kshtsk All commits into 1? Or do you want me to keep some specific stages?

**Comment by kshtsk** [comment]
> @kshtsk All commits into 1? Or do you want me to keep some specific stages?

all in one, since it is initial version, I don't have any tender feelings for the hardening of the first version :-) And it would be good to have it monolithic.

**Comment by kshtsk** [comment]
@susebot run deploy

**Comment by grapheo12** [comment]
Is this somehow error related to this PR:
`teuthology.exceptions.CommandFailedError: Command failed on target-pr-1432-045 with status 1: 'sudo yum -y install ceph-radosgw'`

**Comment by kshtsk** [comment]
@grapheo12 I don't think so, but is nice to see how the scrape parses errors:
https://ceph-ci.suse.de/job/teuthology-deploy-ovh/270/artifact/logs/run/scrape.log/*view*/

**Comment by kshtsk** [comment]
The failure is related to fedora repos accessibility (see report https://ceph-ci.suse.de/job/teuthology-deploy-ovh/270/artifact/logs/run/teuthology-job-3.html):
```
2020-06-26T15:32:46.188 INFO:teuthology.orchestra.run.target-pr-1432-045.stdout:You can remove cached packages by executing 'dnf clean packages'.
2020-06-26T15:32:46.208 INFO:teuthology.orchestra.run.target-pr-1432-045.stderr:Error: Error downloading packages:
2020-06-26T15:32:46.209 INFO:teuthology.orchestra.run.target-pr-1432-045.stderr:  Status code: 503 for https://mirrors.fedoraproject.org/metalink?repo=epel-8&arch=x86_64&infra=genclo&content=centos
2020-06-26T15:32:46.281 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

**Comment by kshtsk** [comment]
I would change the commit title to `results: add scrape after run finished` so it is clear to which teuthology package the scrape is added.

I am fine this PR to go, still @jdurgin review required.

**Comment by grapheo12** [comment]
Apologies from me to drag this for so long :sweat_smile:
